tests/test_bootstrap.py

```diff
@@ -1557,30 +1557,50 @@ async def test_no_base_platforms_loaded_before_recorder(hass: HomeAssistant) ->
         if "recorder" in integrations:
             break
 
-    integrations_or_execs = await loader.async_get_integrations(
+    integrations_or_excs = await loader.async_get_integrations(
         hass, integrations_before_recorder
     )
-    integrations: list[Integration] = []
-    resolve_deps_tasks: list[asyncio.Task[bool]] = []
-    for integration in integrations_or_execs.values():
-        assert not isinstance(integrations_or_execs, Exception)
-        integrations.append(integration)
-        resolve_deps_tasks.append(integration.resolve_dependencies())
-
-    await asyncio.gather(*resolve_deps_tasks)
-    base_platform_py_files = {f"{base_platform}.py" for base_platform in BASE_PLATFORMS}
-    for integration in integrations:
+    integrations: dict[str, Integration] = {}
+    for domain, integration in integrations_or_excs.items():
+        assert not isinstance(integrations_or_excs, Exception)
+        integrations[domain] = integration
+
+    integrations_all_dependencies = await loader.resolve_integrations_dependencies(
+        hass, integrations.values()
+    )
+    all_integrations = integrations.copy()
+    all_integrations.update(
+        (domain, loader.async_get_loaded_integration(hass, domain))
+        for domains in integrations_all_dependencies.values()
+        for domain in domains
+    )
+
+    problems: dict[str, set[str]] = {}
+    for domain in integrations:
         domain_with_base_platforms_deps = BASE_PLATFORMS.intersection(
-            integration.all_dependencies
-        )
-        assert not domain_with_base_platforms_deps, (
-            f"{integration.domain} has base platforms in dependencies: "
-            f"{domain_with_base_platforms_deps}"
+            integrations_all_dependencies[domain]
         )
+        if domain_with_base_platforms_deps:
+            problems[domain] = domain_with_base_platforms_deps
+    assert not problems, f"Integrations have base platforms in dependencies: {problems}"
+
+    base_platform_py_files = {f"{base_platform}.py" for base_platform in BASE_PLATFORMS}
+    exceptions = {
```

Comment:
`exceptions` is not a great name for this allow-list, I suggest to name it something which includes the word "allow" or "ignore".

Also, move it to the top of the function to make it stand out.